### ISSUE TITLE
Falsy error in `array-callback-return`

### ISSUE THREAD

**Issue by kevva** [eslint/eslint#6014]
Returning something inside a `try/catch` block inside `arr.map` will sometimes falsely report an error. The following code should be perfectly valid (since there's an `process.exit()` which should imply that it should return or exit). Currently I have to assign my return to a variable and return it outside the `try/catch`.

```js
arr.map(() => {
    try {
        return require('foo');
    } catch (err) {
        console.error(err);
        process.exit(1);
    }
});
```

Using ESLint version `2.9.0` and the failing rule is `array-callback-return`.

Ref https://github.com/imagemin/imagemin-cli/pull/6#discussion_r61607411.

**Comment by platinumazure** [comment]
I don't know. `process.exit()` is Node-specific. I can't imagine this is a nearly common enough pattern to justify rule logic detecting code specific to Node.

My personal recommendation would be to use `// eslint-disable-line array-callback-return for just this invocation. But perhaps others on the team may have a different opinion.

**Comment by sindresorhus** [comment]
I don't really see the problem with it being Node.js specific. ESLint already have Node.js and browser specific rules. The use of `process.exit()` is a very common thing in CLI apps and workers. Maybe if the ESLint plugin API exposed a method to mark something as unreachable, in similar manner to `markVariableAsUsed()`, we could fix this ourselves. The alternative is having to reimplement this rule in userland, which would be a shame for such a small tweak.

> My personal recommendation would be to use `// eslint-disable-line array-callback-return for just this invocation.

This issue wouldn't have been opened if that was something we wanted. Not interested in littering our code with a myriad of `eslint-disable` statements.

**Comment by platinumazure** [comment]
Just want to go on the record and clarify my position: I don't believe a
rule *that isn't already runtime-specific* should *become runtime-specific*
by handling runtime-specific cases. That was what I meant in my initial
reply.

**Comment by jamestalmage** [comment]
Adding an exception for `process.exit()` does not really limit the utility of this rule when used for code targeted at browsers. Saying it would make the rule "runtime specific" seems to be overstating things a bit.

**Comment by platinumazure** [comment]
Maybe. It's just my personal opinion, not an official pronouncement from
the project. Let's see what the rest of @eslint/eslint-team think.

**Comment by Qix-** [comment]
On the fence about this. Is ESLint a static analyzer or a syntax checker? Other languages support the idea that `exit()`-esque calls are considered end-of-branch. If we already have Node-specific stuff then definitely :+1: for this.

**Comment by SamVerschueren** [comment]
I'm 👍 on this because it exits the execution of the current function (and of the entire program) just like a `throw` statement does. So I believe it's a valid exit path and developers shouldn't write workarounds for the ESLint error, or worse, disable the rule.

**Comment by Qix-** [comment]
Not to be devil's advocate, but:

```javascript
var foo = (Math.random() > 0.5);

function exitIfFoo() {
    if (foo) process.exit(1);
}

arr.map(() => {
    try {
        return require('foo');
    } catch (err) {
        console.error(err);
        exitIfFoo();
    }
});
```

What happens there?

**Comment by sindresorhus** [comment]
@Qix- The exact same as if it was a `throw` instead of `process.exit`, nothing. ESLint is a static analysis tool. Most `process.exit()` calls are inline anyways though.

-

```js
return process.exit(1);
```

👎 I personally don't return stuff that doesn't return anything to be explicit.

**Comment by Qix-** [comment]
@sindresorhus by nothing, what do you mean? That it *would* error requiring a return statement?

---

> I personally don't return stuff that doesn't return anything to be explicit.

I agree that it's not ideal. 'Twas just a workaround. 🍔

**Comment by mikesherov** [comment]
Besides `process.exit()`, what other code patterns should we be lenient about? My concern isn't that this is runtime specific, my concern is that we become the holders of a potentially ever growing list of exceptions.

However, giving plugins the ability to mark a code path as unreachable seems like a nice alternative. @sindresorhus, would that satisfy your requirements?
@eslint/eslint-team thoughts?

**Comment by sindresorhus** [comment]
> by nothing, what do you mean? That ESLint would error requiring a return statement?

@Qix- Yes, ESLint wouldn't be able to detect it.

**Comment by sindresorhus** [comment]
@mikesherov While not preferable, it would solve my problem, and it would be a much better alternative than having to recreate this rule in userland.

**Comment by mikesherov** [comment]
Just don't want to get to a case where we create exceptions for window.close and a large list of other exit functions

**Comment by platinumazure** [comment]
What @mikesherov is saying about a long list of exceptions is exactly why I expressed reluctance to add one for one environment. Sorry for not being more clear on that point.

If we can gather this into a higher level module so that other rules can share that understanding of "what else counts as an exit point?", then I could get on board.

**Comment by nzakas** [comment]
Sorry, late to the party here. We don't want to add `process.exit` as an exception to this rule. Yes, we have Node-specific rules, including one that flags `process.exit`, but this rule is not Node-specific and therefore shouldn't have any Node-specific functionality. 

I think the best approach here is for you to manually disable the rule on that line.

**Comment by mysticatea** [comment]
I have added [node/process-exit-as-throw] rule experimentally to [eslint-plugin-node].
If this rule is turned on, ESLint's code path analysis will address `process.exit()` as `throw`.
This might help this issue.

[node/process-exit-as-throw]: https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/process-exit-as-throw.md
[eslint-plugin-node]: https://github.com/mysticatea/eslint-plugin-node#readme

**Comment by sindresorhus** [comment]
@nzakas Would you be open to giving plugins the ability to mark a code path as unreachable?

**Comment by nzakas** [comment]
@sindresorhus  there's no way it could actually work based on how ESLint works. You'd have to somehow ensure that a plugin rule doing that would run before all other rules that might check unreachable paths.

I think @mysticatea's rule is the best solution here, so closing.
